src/mpl2/src/hier_rtlmp.cpp

```diff
@@ -6638,4 +6241,235 @@ void HierRTLMP::setDebugShowBundledNets(bool show_bundled_nets)
   graphics_->setShowBundledNets(show_bundled_nets);
 }
 
+//////// Pusher ////////
+
+Pusher::Pusher(Cluster* root,
+               odb::dbBlock* block,
+               const std::map<Boundary, Rect>& boundary_to_io_blockage)
+    : root_(root), block_(block)
+{
+  core_ = block_->getCoreArea();
+  dbu_ = block_->getTech()->getDbUnitsPerMicron();
+
+  setIOBlockages(boundary_to_io_blockage);
+}
+
+void Pusher::setIOBlockages(
+    const std::map<Boundary, Rect>& boundary_to_io_blockage)
+{
+  for (const auto& [boundary, box] : boundary_to_io_blockage) {
+    boundary_to_io_blockage_[boundary]
+        = odb::Rect(micronToDbu(box.getX(), dbu_),
+                    micronToDbu(box.getY(), dbu_),
+                    micronToDbu(box.getX() + box.getWidth(), dbu_),
+                    micronToDbu(box.getY() + box.getHeight(), dbu_));
+  }
+}
+
+void Pusher::fetchMacroClusters(Cluster* parent,
+                                std::vector<Cluster*>& macro_clusters)
+{
+  for (Cluster* child : parent->getChildren()) {
+    if (child->getClusterType() == HardMacroCluster) {
+      macro_clusters.push_back(child);
+    } else if (child->getClusterType() == MixedCluster) {
+      fetchMacroClusters(child, macro_clusters);
+    }
+  }
+}
+
+void Pusher::pushMacrosToCoreBoundaries()
+{
+  std::vector<Cluster*> macro_clusters;
+  fetchMacroClusters(root_, macro_clusters);
+
+  for (Cluster* macro_cluster : macro_clusters) {
+    // We don't push macros to the boundaries if that will destroy
+    // the shape of an array of macros.
+    bool horizontal_move_allowed = true;
+    bool vertical_move_allowed = true;
+
+    const int width = micronToDbu(macro_cluster->getWidth(), dbu_);
+    if (width > (core_.dx() / 2)) {
+      horizontal_move_allowed = false;
+    }
+
+    const int height = micronToDbu(macro_cluster->getHeight(), dbu_);
+    if (height > (core_.dy() / 2)) {
+      vertical_move_allowed = false;
+    }
+
+    if (!horizontal_move_allowed && !vertical_move_allowed) {
+      continue;
+    }
+
+    std::map<Boundary, int> boundaries_distance = getDistanceToCloseBoundaries(
+        macro_cluster, vertical_move_allowed, horizontal_move_allowed);
+
+    pushMacroClusterToCoreBoundaries(macro_cluster, boundaries_distance);
+  }
+}
+
+std::map<Boundary, int> Pusher::getDistanceToCloseBoundaries(
+    Cluster* macro_cluster,
+    bool vertical_move_allowed,
+    bool horizontal_move_allowed)
+{
+  std::map<Boundary, int> boundaries_distance;
+
+  std::vector<HardMacro*> hard_macros = macro_cluster->getHardMacros();
+
+  int min_dx_to_core = std::numeric_limits<int>::max();
+  int min_dy_to_core = std::numeric_limits<int>::max();
+
+  for (HardMacro* hard_macro : hard_macros) {
+    hard_macros_.push_back(hard_macro);
+
+    min_dx_to_core = std::min(min_dy_to_core, hard_macro->getWidthDBU());
+    min_dy_to_core = std::min(min_dy_to_core, hard_macro->getHeightDBU());
```

Comment:
The variable name seems to suggest a distance but the value is w/h.  Please explain your thinking.

Reply:
The names should just be w/h as the value, those names weren't a good choice. I also realized that we don't need to iterate the HardMacros within the macro cluster and make all this std::min stuff as we only group macros of the same size (I'll comment this on the code).